pkg/controller/product/product_controller.go

```diff
@@ -0,0 +1,416 @@
+	reconciler, err := newReconciler(mgr)
+	if err != nil {
+		return err
+	}
+
+	return add(mgr, reconciler)
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) (reconcile.Reconciler, error) {
+	apiClientReader, err := common.NewAPIClientReader(mgr)
+	if err != nil {
+		return nil, err
+	}
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(mgr.GetConfig())
+	if err != nil {
+		return nil, err
+	}
+
+	client := mgr.GetClient()
+	scheme := mgr.GetScheme()
+	ctx := context.TODO()
+	recorder := mgr.GetEventRecorderFor(controllerName)
+	return &ReconcileProduct{
+		BaseReconciler: reconcilers.NewBaseReconciler(client, scheme, apiClientReader, ctx, log, discoveryClient, recorder),
+	}, nil
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource Product
+	err = c.Watch(&source.Kind{Type: &capabilitiesv1beta1.Product{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileProduct implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileProduct{}
+
+// ReconcileProduct reconciles a Product object
+type ReconcileProduct struct {
+	*reconcilers.BaseReconciler
+}
+
+// Reconcile reads that state of the cluster for a Product object and makes changes based on the state read
+// and what is in the Product.Spec
+func (r *ReconcileProduct) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := r.Logger().WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconcile Product", "Operator version", version.Version)
+
+	// Fetch the Product instance
+	product := &capabilitiesv1beta1.Product{}
+	err := r.Client().Get(r.Context(), request.NamespacedName, product)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			reqLogger.Info("resource not found. Ignoring since object must have been deleted")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if reqLogger.V(1).Enabled() {
+		jsonData, err := json.MarshalIndent(product, "", "  ")
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		reqLogger.V(1).Info(string(jsonData))
+	}
+
+	// Ignore deleted Products, this can happen when foregroundDeletion is enabled
+	// https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion
+	if product.DeletionTimestamp != nil {
```

Comment:
Can this also happen for other CRs like APIManager?

Reply:
Sure! if you delete APIManager with `foregrounddeletion` enabled you will get reconciliation loop with `DeletionTimestamp` set and if not checked, it will try to reconciliate spec as always. It can also happen if finalizers are used.

Reply:
Ok. We should think on adding it to the others then in the future to avoid this possible race condition